### PR TITLE
OCaml 4.08 compatibility

### LINKED ISSUE #304: Odoc fails to compile with OCaml 4.08
The constructors of `Types.signature_item` have been extended with a `visibility` flag introduced in ocaml/ocaml#2147. On first glance it looks like we should just filter out items that are `Hidden`, but I'd like to double check that.

### PR DESCRIPTION
**Pull request by jonludlam**
Further work needs to be done to support the new typesubst/modsubst constructors.

Fixes #304

### CODE REVIEW

File: src/loader/doc_attr.ml [hunk 1]
```diff
@@ -46,9 +46,15 @@ let attached parent attrs =
...
     function
+#if OCAML_MAJOR = 4 && OCAML_MINOR >= 08
+    | {Parsetree.attr_name = { Location.txt =
+          ("doc" | "ocaml.doc"); loc = _loc}; attr_payload; _ } :: rest -> begin
+        match load_payload attr_payload with
+#else
     | ({Location.txt =
           ("doc" | "ocaml.doc"); loc = _loc}, payload) :: rest -> begin
         match load_payload payload with
+#endif
```
Comment (trefis) [inline src/loader/doc_attr.ml]: You could limit the `#if .. #else .. #endif` to the pattern part if you make sure to use the same variable names on both sides.

File: src/loader/doc_attr.ml [hunk 2]
```diff
@@ -93,9 +99,15 @@ let standalone parent
...
   function
+#if OCAML_MAJOR = 4 && OCAML_MINOR >= 08
+  | { attr_name = { Location.txt =
+        ("text" | "ocaml.text"); loc = _loc}; attr_payload; _ } -> begin
+      match load_payload attr_payload with
+#else
   | ({Location.txt =
         ("text" | "ocaml.text"); loc = _loc}, payload) -> begin
       match load_payload payload with
+#endif
```
Comment (trefis) [inline src/loader/doc_attr.ml]: Same as above.

File: src/loader/cmt.ml [hunk 3]
```diff
@@ -78,6 +78,10 @@ let rec read_pattern env parent doc pat =
         read_pattern env parent doc pat
     | Tpat_lazy pat ->
         read_pattern env parent doc pat
+#if OCAML_MAJOR = 4 && OCAML_MINOR >= 08
+    | Tpat_exception _ ->
+      []
```
Comment (trefis) [inline src/loader/cmt.ml]: That doesn't look right. I think you should recurse in that case.

Comment (jonludlam) [reply src/loader/cmt.ml]: Yes, I think you're right.

File: src/loader/doc_attr.ml (hunk 1 above)
Comment (jonludlam) [reply src/loader/doc_attr.ml]: True, that'll probably look nicer. Thanks!

**Comment by aantron** [comment]
Thanks @jonludlam and @trefis. I'll merge once the CI is done. Looks stuck for now :/ I'll add a 4.08 build to Travis right after merging (or push into this PR if the current CI build still hasn't started by then).

**Comment by jonludlam** [comment]
The 4.08 travis build might be tricky until `ocaml-migrate-parsetree` is available. We'd have to remove the bisect_ppx dependency.

**Comment by aantron** [comment]
Looks like we also transitively depend on Ocamlbuild, which needs a release (https://github.com/ocaml/ocamlbuild/issues/294).

**Comment by jonludlam** [comment]
True - though there at least pinning master works.

**Comment by aantron** [comment]
@jonludlam or @trefis, would you mind opening a follow-up issue about "proper" support for the new 4.08 features?

**Comment by avsm** [comment]
I've opened up #350 to track the new features in 4.08

**Pull request merged**